ICU-22119 Add lw=phrase for Korean using line_*_phrase_cj

jungshik opened the pull request:
brkitr/ko.txt is created to use line_*_.cj.txt for both
lw=phrase and lw != phrase cases for Korean. This is the simplest
way to fix ICU-22119 taking advantage of the fact that ICU
does not have a Korean dictionary so we don't have to worry about
adding the list of Korean particles to keep them attached to the
preceeding word.

The downside is that it only works when the locale is ko or ja while
it should work in any locale. Another is it makes ICU deviate from
CSS3 by using the same CJ (conditonal Japanese) rules for Korean as
well. However, CSS3 spec is wrong on that point and should be changed.
See https://unicode-org.atlassian.net/browse/CLDR-4931 .

<!--
Thank you for your pull request!

Please see http://site.icu-project.org/processes/contribute for general
information on contributing to ICU.

You will be automatically asked to sign the contributors license agreement (CLA) before the PR is accepted.
- sign: https://cla-assistant.io/unicode-org/icu
- license: http://www.unicode.org/copyright.html
-->

##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22119
- [x] Required: The PR title must be prefixed with a JIRA Issue number. <!-- For example: "ICU-1234 Fix xyz" -->
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. <!-- For example: "ICU-1234 Fix xyz" -->
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

jungshik commented:
This is not yet for review-ready. This is to show how the 2nd approach listed in ICU-22119 is implemented. Once we reach an agreement as to which approach to take for ICU-22119, I'll update either this one or other PR.

FrankYFTang commented:
@allensu05

markusicu commented:
Hi @jungshik, in the ICU meeting last week we agreed on your option 2. Do you have time to get this ready in the next few days, so that it can be reviewed and merged by the end of next week = ICU 72 code freeze?

- Please add unit tests.
- I think the data file will need a corresponding .xml file from which it should be generated, that is, a ko.xml that is probably initially a copy of https://github.com/unicode-org/icu/blob/main/icu4c/source/data/xml/brkitr/ja.xml
- Should the currently-identical lists of particles ("extensions") be pushed into the root data file? Or not useful because they should eventually diverge?

pedberg-icu commented on icu4c/source/data/brkitr/ko.txt:
```diff
@@ -0,0 +1,242 @@
+﻿// © 2016 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+// Generated using tools/cldr/cldr-to-icu/build-icu-data.xml
```
As it says, this file is actually supposed to be generated, from xml data in data/xml/brkitr/. Take a look at data/xml/brkitr/ja.xml and add something similar for ko.

jungshik replied:
Thank you, Peter.   I added ko.xml . Should I delete ko.txt change, then?

jungshik replied:
ooops.  perhaps, I still need to keep ko.txt (even though it's not generated by the tool). Or, should I build a CLDR tool and generate from the xml file? The result should be identical (modulo space/format, etc).

jungshik replied:
> Please add unit tests.

I added several lines of test to rbbi/RBBITest/TestExtended . Is it sufficient?

> Should the currently-identical lists of particles ("extensions") be pushed into the root data file? Or not useful because they should eventually diverge?

It turned out that it's not necessary to duplicate the list. In the long run, I'd argue that the list should be moved to root, but that's for another day. 

I added a comment about Korean particle list for the future.

pedberg-icu replied:
> Thank you, Peter. I added ko.xml . Should I delete ko.txt change, then?

No need to delete the ko.txt change now; the change to ko.xml will just prevent your ko.txt change from being eliminated next time there is a CLDR-ICU integration.

jungshik replied:
Got it. Thank you.

pedberg-icu commented:
Please add tests in test/testdata/rbbitst.txt that show the different between ko and ko@lw=phrase, thanks.

pedberg-icu reviewed (changes requested):
Please add data/xml/brkitr/ko.xml (as mentioned) and appropriate tests in test/testdata/rbbitst.txt

jungshik commented:
> Please add tests in test/testdata/rbbitst.txt that show the different between ko and ko@lw=phrase, thanks.

I added ko@lw=phrase case, but didn't think of adding tests for ko. I'm taking care of that now.

jungshik commented:
Sorry that I haven't addressed review comment early enough. I think I addressed them all. Please, take another look .

@pedberg-icu @markusicu (I didn't mean to drop @markusicu and others from the reviewer's list, but re-requesting reviews had that effect).

jungshik commented:
> Please add data/xml/brkitr/ko.xml (as mentioned) and appropriate tests in test/testdata/rbbitst.txt

Both have been done.

aheninger commented on icu4c/source/test/testdata/rbbitst.txt:
```diff
@@ -1981,6 +1981,57 @@ Bangkok)•</data>
 #これらの連絡先はデバイスをロック解除しなくても表示され -> これらの▁連絡先は▁デバイスを▁ロック▁解除しなくても▁表示され
 <data>•\u3053\u308C\u3089\u306E•\u9023\u7D61\u5148\u306F•\u30C7\u30D0\u30A4\u30B9\u3092•\u30ED\u30C3\u30AF•\u89E3\u9664\u3057\u306A\u304F\u3066\u3082•\u8868\u793A\u3055\u308C•</data>
 
+# Test the differences in ko with or without lw=phrase.
```
There's a second copy of rbbitst.txt in ICU4J.
The two are supposed to always be identical; changes in one that aren't in the other are at risk of being overwritten by other updates.

jungshik replied:
Thank you, Andy !   Should I fold the corresponding Java (perhaps 1 line code change) into this PR?  Oh. You mentioned it below. I'll do that and update the PR.

aheninger commented on icu4c/source/common/brkiter.cpp:
```diff
@@ -439,8 +439,8 @@ BreakIterator::makeInstance(const Locale& loc, int32_t kind, UErrorCode& status)
-            // lw=phrase is only supported in Japanese.
-            if (uprv_strcmp(loc.getLanguage(), "ja") == 0) {
+            // lw=phrase is only supported in Japanese and Korean
+            if (uprv_strcmp(loc.getLanguage(), "ja") == 0 || uprv_strcmp(loc.getLanguage(), "ko") == 0) {
```
This needs to be done for ICU4J as well.

The new data will end up on the J side  at some point anyway, independently of this PR,  the first time anyone refreshes J data from ICU4C. Best to keep everything in sync, and do Java at the same time as C.

jungshik replied:
Done in the latest PR.

jungshik commented:
```
Testsuite: com.ibm.icu.dev.test.rbbi.RBBITestExtended
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.086 sec

Testcase: TestExtended took 0.081 sec
        Caused an ERROR
java.util.MissingResourceException: Can't find resource for bundle com.ibm.icu.impl.ICUResourceBundleImpl$ResourceTable, key 2
java.util.MissingResourceException: java.util.MissingResourceException: Can't find resource for bundle com.ibm.icu.impl.ICUResourceBundleImpl$ResourceTable, key 2
        at com.ibm.icu.text.BreakIteratorFactory.createBreakInstance(BreakIteratorFactory.java:156)
        at com.ibm.icu.text.BreakIteratorFactory.createBreakIterator(BreakIteratorFactory.java:72)
        at com.ibm.icu.text.BreakIterator.getBreakInstance(BreakIterator.java:902)
        at com.ibm.icu.text.BreakIterator.getLineInstance(BreakIterator.java:669)
        at com.ibm.icu.dev.test.rbbi.RBBITestExtended.TestExtended(RBBITestExtended.java:170)
```

I suspect that the above error is caused by  com/ibm/icu/impl/data/icudt72b/brkitr/ko.res missing in icudata.jar . 

What should I do to get it added to icudata.jar ?

pedberg-icu commented:
> I suspect that the above error is caused by  com/ibm/icu/impl/data/icudt72b/brkitr/ko.res missing in icudata.jar .
>
> What should I do to get it added to icudata.jar ?

Hi Jungshik, you can do that using steps 12-12a-12b in https://github.com/unicode-org/icu/blob/main/icu4c/source/data/cldr-icu-readme.txt, assuming that (1) you already have done a full build of ICU4C so that the data build tools are already created, and (2) you have export symbols defined as in steps 1a-1c in thay readme.txt doc.

I noticed one other issue for generating brkitr/ko.txt from xml/brkitr/ko.xml: Besides the things I already mentioned, "ko" needs to be added to the appropriate line in tools/cldr/cldr-to-icu/build-icu-data.xml: https://github.com/unicode-org/icu/blob/main/tools/cldr/cldr-to-icu/build-icu-data.xml#L359

jungshik commented:
Peter, can you take another look? Thanks !


After a few attempts of rebuilding ICU and icudata.jar, I couldn't resolve one test failure associated with the following line in the test file. Mysteriously, ICU4C is happy, but ICU4J is not. 

```
<data>•"님의 •침묵"을 •읽고•</data>
```

So, I commented it out and added a note about it. With that, ICU4J is happy.

jungshik commented:
Andy, can you also take another look? Thank you

aheninger commented:
> Andy, can you also take another look? 
The changes look plausible to me.

It might be nice to confirm that rebuilding ICU data from CLDR really does generate the expected ko.txt file

pedberg-icu commented:
> It might be nice to confirm that rebuilding ICU data from CLDR really does generate the expected ko.txt file

I will probably do another CLDR-ICU integration this week and as part of that can either confirm that it does, or fix the problems if it does not. So I think that should not hold up this PR.

pedberg-icu commented:
> After a few attempts of rebuilding ICU and icudata.jar, I couldn't resolve one test failure associated with the following line in the test file. Mysteriously, ICU4C is happy, but ICU4J is not.
> ```
> <data>•"님의 •침묵"을 •읽고•</data>
> ```
> So, I commented it out and added a note about it. With that, ICU4J is happy.

Hmm, I don't see that line commented out in ICU4J. And there is still an ICU4J test failure.

jungshik commented:
> Hmm, I don't see that line commented out in ICU4J. And there is still an ICU4J test failure

The line was commented out in both ICU4J and ICU4C . See 
<img width="724" alt="image" src="https://user-images.githubusercontent.com/8578343/193743429-26fdee81-14a6-4571-8882-7d4534e11a0e.png"> :-) 

Not sure why lstm-icu4j fails while java[8,11,16]-icu4j bots pass. Before I commented that line out, java[8,11,16]-icu4j bots all failed.

aheninger commented:
> Not sure why lstm-icu4j fails while java[8,11,16]-icu4j bots pass. Before I commented that line out, java[8,11,16]-icu4j bots all failed.

Commenting the test line out for both C and J was the right thing to do. There is no mechanism in place for maintaining differences between the C and J copies of the test file.

Jungshik's branch passes all Java tests for me, locally. (For another random testing data point)

```
$ java -version
openjdk version "11.0.16" 2022-07-19
OpenJDK Runtime Environment (build 11.0.16+8-post-Ubuntu-0ubuntu122.04)
```

jungshik commented:
> Commenting the test line out for both C and J was the right thing to do.
Yeah, I disabled the test in question in both ICU4C and ICU4J. (actually, I just copied the former to the latter). 

As for lstm-icu4j,  what failed is not what I edited ( com.ibm.icu.dev.test.rbbi.RBBITestExtended is skipped in lstm-icu4j)  but com.ibm.icu.dev.test.util.Trie2Test failed. 

```
  [junit] Running com.ibm.icu.dev.test.util.Trie2Test
    [junit] Tests run: 4, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.571 sec
```

```
 [junit] Running com.ibm.icu.dev.test.rbbi.RBBITestExtended
    [junit] Tests run: 1, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.002 sec
```

jungshik commented:
Below is details about the failed test on lstm-icu4j  bot.  It seems flaky because my change has nothing to do with trie2. 

How can I try that locally? From the name, it seems to be ICU4J with LSTM enabled for segmentation (Burmese and Khmer? ).   

```

Testcase: TestTrie2API took 0.047 sec
Testcase: TestRanges took 1.493 sec
	FAILED
: wrong value for set-single-value of d800: expected 5, got 3
junit.framework.AssertionFailedError: : wrong value for set-single-value of d800: expected 5, got 3
	at com.ibm.icu.dev.test.AbstractTestLog.errln(AbstractTestLog.java:80)
	at com.ibm.icu.dev.test.TestFmwk.handleAssert(TestFmwk.java:642)
	at com.ibm.icu.dev.test.TestFmwk.handleAssert(TestFmwk.java:623)
	at com.ibm.icu.dev.test.TestFmwk.assertEquals(TestFmwk.java:538)
	at com.ibm.icu.dev.test.util.Trie2Test.trieGettersTest(Trie2Test.java:649)
	at com.ibm.icu.dev.test.util.Trie2Test.checkTrieRanges(Trie2Test.java:738)
	at com.ibm.icu.dev.test.util.Trie2Test.TestRanges(Trie2Test.java:758)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
```

jungshik commented:
@pedberg-icu and @aheninger :  do you have any idea as to how to make LSTM Java bot happy?  As I wrote above, the failing test has nothing to do with my change.

aheninger commented:
> @pedberg-icu and @aheninger : do you have any idea as to how to make LSTM Java bot happy? As I wrote above, the failing test has nothing to do with my change.

I clicked the "Re-run this job" thingy for the failing test, and it now passes. Some kind of glitch, somewhere.

The lstm-icu4j test is doing a complete rebuild of the ICU4J data from ICU4C, so it might well be turning up data-related failures that aren't caught by other CI testing. From the test log,

```
Run cd icu4c/source;
  cd icu4c/source;
  ICU_DATA_BUILDTOOL_OPTS=--include_uni_core_data ICU_DATA_FILTER_FILE=../../.github/lstm_for_th_my.json ./runConfigureICU --enable-debug --disable-release Linux -disable-layoutex;
  make clean;
  make -j[2](https://github.com/unicode-org/icu/actions/runs/3156996537/jobs/5274277463#step:5:2) ICU[4](https://github.com/unicode-org/icu/actions/runs/3156996537/jobs/5274277463#step:5:4)J_ROOT=../../../icu4j icu4j-data-install;
  cd ../..
```

jungshik commented:
Thank you, @aheninger !   

What'd be next step for me to take to get this submitted?  Do you have any changes for me to make?  The same question goes to @pedberg-icu .

Thanks !

jungshik commented:
@pedberg-icu , @aheninger @FrankYFTang :  can you take a look?

@markusicu : I've done everything requested. Pls, drop 'wait-on-author' label. 

Thank you

pedberg-icu commented:
This looks good to me now. One of the commit message is missing the ICU-22119 prefix. This should get fixed either when squashing, or by amending the commit message.

FrankYFTang commented:
@allensu05 are you ok with this PR?

allensu05 commented:
> @allensu05 are you ok with this PR?

It looks good to me.

jungshik commented:
@pedberg-icu thank you for the approval. Due to a recent IANA tzdb update (to 2022f), there was a merge conflict to resolve. I just resolved it. Can you approve again?  

I can't merge myself (I'm not authorized to). After approving it, could you squash and merge?  Thank you

jungshik commented:
> This looks good to me now. One of the commit message is missing the [ICU-22119](https://unicode-org.atlassian.net/browse/ICU-22119) prefix. This should get fixed either when squashing, or by amending the commit message.

Thank you for the approval. 

The commit message was fixed to have ICU-22119.  That change and icutzdata.jar update require a new approval. Can you approve again ?  Thank you

jungshik commented:
> > This looks good to me now. One of the commit message is missing the [ICU-22119](https://unicode-org.atlassian.net/browse/ICU-22119) prefix. This should get fixed either when squashing, or by amending the commit message.
> 
> Thank you for the approval.
> 
> The commit message was fixed to have [ICU-22119](https://unicode-org.atlassian.net/browse/ICU-22119). That change and icutzdata.jar update require a new approval. Can you approve again ? Thank you

I also squashed to a single commit.

pedberg-icu reviewed (approved):
looks good now, thanks.

pedberg-icu commented:
I can merge once tests finish running.

jungshik commented:
> I can merge once tests finish running.

Thank you !  I wonder why this is still pending a while after all GHA CI's passed . Hope it'll get updated soon. 

> CI Expected — Waiting for status to be reported

pedberg-icu commented:
/azp run CI-Exhaustive

pedberg-icu commented:
> I wonder why this is still pending a while after all GHA CI's passed . Hope it'll get updated soon.

I started the exhaustive tests for this (probably a good idea anyway), maybe when those complete it will jog the overall CI result into showing complete.

pedberg-icu commented:
> I started the exhaustive tests for this (probably a good idea anyway), maybe when those complete it will jog the overall CI result into showing complete.

Well, that did not work. I have a requested a re-run of the CI tests, hopefully after those complete the overall status will show as completed.

jungshik commented:
> > I started the exhaustive tests for this (probably a good idea anyway), maybe when those complete it will jog the overall CI result into showing complete.
> 
> Well, that did not work. I have a requested a re-run of the CI tests, hopefully after those complete the overall status will show as completed.

Hmm, even that didn't work.
Strange....

jungshik commented:
It's been a few days and the overall ci check is still pending even though every bot succeeded. How can it be unblocked?

aheninger commented:
I just pushed the "Rerun all Jobs" button. We will see what happens in a little bit ...

markusicu commented:
/azp run
